src/AxisSets.jl

```diff
@@ -1,5 +1,218 @@
 module AxisSets
 
-# Write your package code here.
+using AxisKeys
+using Impute
+using IterTools
+using NamedDims
+using OrderedCollections
+using ReadOnlyArrays
+using Tables
+
+using Impute:
+    Filter,
+    Imputor,
+    Validator
+
+# Short hand type for complicated union of nested Keyed or NamedDims arrays
+XArray{L, T, N} = Union{NamedDimsArray{L,T,N,<:KeyedArray}, KeyedArray{T,N,<:NamedDimsArray}}
+
+struct Dataset{T<:XArray}
+    dims::OrderedSet{Symbol}
+    data::LittleDict{Symbol, T}
+end
+
+function Dataset(dims::Symbol...; kwargs...)
+    # If we didn't get any kwargs then just create an empty Dataset
+    isempty(kwargs) && return Dataset(OrderedSet(dims), LittleDict{Symbol, XArray}())
+
+    # Checking the eltype doesn't always work if the `NamedDimsArray` covers different dimensions
+    if all(v -> isa(v, XArray), values(kwargs))
+        # If we were passed a valid XArray then just construct our LittleDict
+        data = LittleDict{Symbol, XArray}(kwargs...)
+    else
+        # Otherwise we need to assume that we're operating over a collection of tables
+        data = LittleDict{Symbol, XArray}()
+
+        # Might be a simpler way to do this if we have a `merge` function or something
+        for (name, table) in kwargs
+            @show typeof(table)
+            # Reuse the table constructor and extra the data
+            tmp = Dataset(table, dims...).data
+
+            if length(tmp) == 1
+                data[name] = first(values(tmp))
+            elseif length(tmp) > 1
+                for (k, v) in tmp
+                    data[Symbol(name, "_", k)] = v
```

Comment:
this looks like it only supports one level of nesting?

It would be nice if could support arbitrary nested Tuples/NamedTuples as that's typically how our features are constructed

Reply:
Yeah, I'm just not sure if it's worth the extra complexity. It's currently unclear how necessary that nesting is, and we don't seem to do a lot of data manipulation operations based around that nesting. We could probably use tuple keys and overload `getproperty` to interface to mimic that organization though.

Reply:
Hmm, actually we could also just overload `getindex` for `Tuple` and `NamedTuple` keys, so that you could do things like `ds[group=:train, val=Load] -> KeyedArray`, `ds[:, Temperature] -> Dataset`, `ds[:predict, :] -> Dataset`?